File: ta_lib/strategies/signal/src/tsi_flip.rs
use base::prelude::*;
use core::prelude::*;
use momentum::tsi;

const TSI_ZERO_LINE: f32 = 0.0;

pub struct TSIFlipSignal {
    smooth_type: Smooth,
    fast_period: usize,
    slow_period: usize,
}

impl TSIFlipSignal {
    pub fn new(smooth_type: Smooth, fast_period: f32, slow_period: f32) -> Self {
        Self {
            smooth_type,
            fast_period: fast_period as usize,
            slow_period: slow_period as usize,
        }
    }
}

impl Signal for TSIFlipSignal {
    fn lookback(&self) -> usize {
        std::cmp::max(self.fast_period, self.slow_period)
    }

    fn generate(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let tsi = tsi(
            &data.close,
            self.smooth_type,
            self.slow_period,
            self.fast_period,
        );

        (
            tsi.cross_over(&TSI_ZERO_LINE),
            tsi.cross_under(&TSI_ZERO_LINE),
        )
    }
}

File: strategy/generator/signal/tsi_flip.py
from dataclasses import dataclass

from core.models.parameter import (
    Parameter,
    StaticParameter,
)
from core.models.smooth import Smooth

from .base import Signal, SignalType


@dataclass(frozen=True)
class TsiFlipSignal(Signal):
    type: SignalType = SignalType.TsiFlip
    smooth_type: Parameter = StaticParameter(Smooth.EMA)
    fast_period: Parameter = StaticParameter(13.0)
    slow_period: Parameter = StaticParameter(25.0)
